src/managers/backup.py

```diff
@@ -5,34 +5,43 @@
 """Helpers for managing backups."""
 
 import logging
+from datetime import datetime
+from io import StringIO
 
 import boto3
+import httpx
+import yaml
 from botocore import loaders, regions
 from botocore.exceptions import ClientError
 from mypy_boto3_s3.service_resource import Bucket
+from rich.console import Console
+from rich.table import Table
 
-from core.stubs import S3ConnectionInfo
+from core.cluster import ClusterState
+from core.stubs import BackupMetadata, S3ConnectionInfo
+from literals import ADMIN_SERVER_PORT
 
 logger = logging.getLogger(__name__)
 
 
 class BackupManager:
     """Manager for all things backup-related."""
 
-    def __init__(self, s3_parameters: S3ConnectionInfo) -> None:
-        self.s3_parameters = s3_parameters
+    def __init__(self, state: ClusterState) -> None:
+        self.state = state
 
     @property
     def bucket(self) -> Bucket:
         """S3 bucket to read from and write to."""
+        s3_parameters = self.state.cluster.s3_credentials
         s3 = boto3.resource(
             "s3",
-            aws_access_key_id=self.s3_parameters["access-key"],
-            aws_secret_access_key=self.s3_parameters["secret-key"],
-            region_name=self.s3_parameters["region"] if self.s3_parameters["region"] else None,
-            endpoint_url=self._construct_endpoint(self.s3_parameters),
+            aws_access_key_id=s3_parameters["access-key"],
+            aws_secret_access_key=s3_parameters["secret-key"],
+            region_name=s3_parameters["region"] if s3_parameters["region"] else None,
```

Comment:
**nitpick:** `s3_parameters.get("region", None)`

Reply:
Actually, we want to replace the empty string with `None` as well. With `""` we get an exception

Reply:
That's fine, doing `{"foo": ""}.get("foo", None)` will evaluate `None` also